File: week3/code/get_TreeHeight.py
#!/usr/bin/env pythons

""" Calculates tree heights given distance of each tree 
from its base and angle to its top, using  the trigonometric formula. 
Outputs to csv called INPUT_FILE_NAME_treeheights_py.csv"""


__appname__ = 'get_TreeHeight.py'
__author__ = 'Elliott Parnell (elliott.parnell22@imperial.ac.uk), Jooyoung Ser (jooyoung.ser19@imperial.ac.uk), Anqi Wang (anqi.wang22@imperial.ac.uk), Linke Feng (L.feng22@imperial.ac.uk)'
__version__ = '0.0.1'

## Imports ##
import numpy as np
import os
import csv
import sys

## Functions ##
def TreeHeight(degrees, distance):
    """Calculate tree heights"""
    radians = degrees * np.pi / 180
    height = distance * np.tan(radians)
    return height  

# Data loaded from command line 
if len(sys.argv) != 2:  # wrong input, exit directly
    print("One(just one) .csv file needed.")
    sys.exit("Please give right input. Try again.") 
    
else:
    with open(sys.argv[1], 'r') as a: 
        data = []    # setup an empty list for data storation
        
        for j in csv.reader(a): 
            data.append(j) 
        data.remove(data[0]) # remove header
            
        for i in data:
            i[1] = float(i[1])
            i[2] = float(i[2])
            i.append(TreeHeight(i[2], i[1])) # save height output to 3rd col
       
       #Write csv results
        with open("../results/"+ os.path.basename(os.path.splitext(sys.argv[1])[0]) + "_treeheights_py.csv", 'w') as b:   
            csv.writer(b).writerow(['Species', 'Distance.m', 'Angle.degrees', 'Tree.Height.m']) # header  ordered
            for t in data:
                csv.writer(b).writerow(t) 

              

File: week3/code/get_TreeHeight.R
#!/usr/bin/env R

# Title: get_TreeHeight.R
# Author details: Jooyoung Ser (jooyoung.ser19@imperial.ac.uk), Elliott Parnell (elliott.parnell22@imperial.ac.uk), Anqi Wang (anqi.wang22@imperial.ac.uk), Linke Feng (L.feng22@imperial.ac.uk)
# Date: Nov 2022
# Script and data info:
# This function calculates heights of trees given distance of each tree 
# from its base and angle to its top, using  the trigonometric formula 
# height = distance * tan(radians)
# Arguments: 1 or 0
# args: csv; degrees: the angle of elevation of tree; distance: the distance from base of tree
# OUTPUT
# The heights of the tree, same units as "distance"


rm(list = ls())

# Function: calculate tree heights
TreeHeight <- function(degrees, distance) {
    radians <- degrees * pi / 180
    height <- distance * tan(radians)
    return (height)
}

# Read data from commandline
args<-commandArgs(trailingOnly=TRUE)

# if no input file, exit directly
if (length(args) != 1) { 
  stop("One(just one) .csv file needed.", call.=F)
} else {
  TreeData <- read.csv(args)
}

TreeData <- read.csv(file = args[1])

# Add a new column and calculate the tree height 
TreeData$Tree.Height.m <- TreeHeight(TreeData$Angle.degrees, TreeData$Distance.m)

# Create a csv output 
write.csv(TreeData, paste("../results/", tools::file_path_sans_ext(basename(args)), "_treeheights_R.csv",sep = ""), row.names = F)

